blocks/xterm/example.cc

```diff
@@ -0,0 +1,104 @@
+#include <iostream>
+#include <sstream>
+#include <thread>
+
+#include "progress.h"
+#include "vt100.h"
+
+inline void wait() { std::this_thread::sleep_for(std::chrono::seconds(1)); }
+
+int main() {
+  {
+    // Basic VT100 demo.
+    using namespace current::vt100;
+
+    std::cout << "Default, " << bold << "bold" << off(bold) << ", " << dim << "dim" << off(dim) << ", done."
+              << std::endl;
+    std::cout << "Default, " << italic << "italic" << off(italic) << '.' << std::endl;
+    std::cout << "Default, " << underlined << "underlined" << off(underlined) << '.' << std::endl;
+    std::cout << "Default, " << bold << italic << underlined << "bold & italic & underlined" << reset << '.'
+              << std::endl;
+
+    std::ostringstream oss;
+    oss << default_color << ", ";
+    std::string const sep = oss.str();
+
+    std::cout << "Default, " << black << "black" << sep << dark_gray << "dark gray" << sep << bright_gray
+              << "bright gray" << sep << white << "white" << sep << "done." << std::endl;
+    std::cout << "Default [ normal ], " << red << "red" << sep << green << "green" << sep << blue << "blue" << sep
+              << yellow << "yellow" << sep << magenta << "magenta" << sep << cyan << "cyan" << sep << "done."
+              << std::endl;
+    std::cout << "Default [ bright ], " << bright_red << "red" << sep << bright_green << "green" << sep << bright_blue
+              << "blue" << sep << bright_yellow << "yellow" << sep << bright_magenta << "magenta" << sep << bright_cyan
+              << "cyan" << sep << "done." << std::endl;
+
+    std::cout << "Default, " << background(red) << blue << bold << "blue on red " << reverse << " reversed "
+              << off(reverse) << " back to blue on red" << reset << '.' << std::endl;
+
+    std::cout << "Default, " << strikeout << "no " << red << "red" << bold << " bold" << italic << " italic"
+              << off(italic) << off(bold) << default_color << " style" << off(strikeout) << '.' << std::endl;
+  }
+
+  {
+    // Smoke test.
+    std::cout << "Test1> ";
+    {
+      current::ProgressLine progress;
+      progress << "Hello,";
+      wait();
+      progress << "Hello, "
+               << "world.";
+      wait();
+      progress << "TTYL!";
+      wait();
+    }
+    std::cout << "Done." << std::endl;
+  }
+  {
+    // Cyrillic.
+    std::cout << "Test2> ";
+    {
+      current::ProgressLine progress;
+      progress << u8"Привет...";
+      wait();
+      progress << u8"Йоу" << '!';
+      wait();
+      progress << u8"Йоу" << ", world!";
+      wait();
+    }
+    std::cout << "Done." << std::endl;
+  }
+  {
+    // Variable length characters.
+    // NOTE(dkorolev): This fails on my Linux. :/ Prints `Test3> tttesting ... OK`, then `Test3> ttDone.` at the end.
```

Comment:
Have a look: https://github.com/sindresorhus/string-width/blob/master/index.js

Reply:
The issue is with Unicode in this case, not control characters. The latter are safely ignored by themselves. (And even UTF8 seems to work just fine with Cyrillic, just not with Japanese.)

Reply:
To elaborate: JavaScript strings are "one index one Unicode character", while in Current it's UTF8.

Reply:
I wonder why it won't work though. There's a notion of code points which are 1 to 2 UTF-16 characters, and 1 to 4 UTF-8 characters. Does Current consider that?

Reply:
Current deals with UTF8, and this functionality (`UTF8StringLength`) is extensively tested and used in other places (FriendlyData).

Reply:
I'm afraid this has to do with the behavior of the terminal (wrt how many `\b`-s does it need), not with how string length is computed.

Thus: a "frontend" issue which I could not care less about :)

Reply:
> That's good. Tried Japanese in its tests? I think I only saw Cyrillic.

[duh](https://github.com/C5T/Current/pull/830/files#diff-31bf74dccc14f47f7f925c6ceb14aa14R79)

> Of course, you haven't unlearned how to behave like a d*ck. ;)

You'd like this book.

![image](https://user-images.githubusercontent.com/2159447/42198754-daf2ab6c-7e3e-11e8-8da1-f8a72ec2af4d.png)

Reply:
In the UTF8StringLength tests, not the terminal progress tests. There's just a smoke with latin and cyrillic, or I'm missing something.

Reply:
Did you click the "duh" link above?

Reply:
Yes I did, it points to a few lines below this comment:
<img width="826" alt="screen shot 2018-07-03 at 10 47 17 am" src="https://user-images.githubusercontent.com/498274/42236071-9f51f692-7eae-11e8-95de-9f78e1e2b388.png">

And here's what I'm talking about:
https://github.com/C5T/Current/blob/d38e4af8c876d92be948a89c22310231981f14db/bricks/strings/test.cc#L947-L950
<img width="618" alt="screen shot 2018-07-03 at 10 48 00 am" src="https://user-images.githubusercontent.com/498274/42236085-a7ef88e6-7eae-11e8-88f3-91086d239d52.png">

Reply:
So, your question was "Are Japanese characters tested wrt the `UTF8StringLength()` code?". Now I get it.

Maybe not in this repo, but we use UTF8 in prod, and I've double-checked the logic of UTF8 symbol boundaries, which makes me rather confident that the issue is not here.